netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -280,10 +283,29 @@ protected boolean isGracefulShutdownComplete() {
    * the creation request is queued.
    */
   private void createStream(CreateStreamCommand command, final ChannelPromise promise) {
-    final int streamId = getAndIncrementNextStreamId();
+    // Get the stream ID for the new stream.
+    final int streamId;
+    try {
+      streamId = getAndIncrementNextStreamId();
+    } catch (StatusException e) {
+      // Stream IDs have been exhausted for this connection. Fail the promise immediately.
+      promise.setFailure(e);
+
+      // Initiate a graceful shutdown if we haven't already.
+      if (!connection().goAwaySent()) {
+        logger.fine("Stream IDs have been exhausted for this connection. "
+                + "Initiating graceful shutdown of the connection.");
+        try {
+          super.close(ctx, ctx.newPromise());
+        } catch (Exception ce) {
+          // Promise was already failed. Just absorb.
```

Comment:
Except that our method signature should simply "throws Exception". The only reason we have this exception is because Netty's close() has "throws Exception", and I see no reason why we shouldn't propagate.